File: app/src/main/java/com/dhy/apiholder/demo/MainActivity.java
package com.dhy.apiholder.demo;

import android.database.Observable;
import android.os.ResultReceiver;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;

import com.dhy.apiholder.ApiHolderUtil;
import com.dhy.apiholder.PartApi;

import java.util.ArrayList;
import java.util.List;

import okhttp3.ResponseBody;
import retrofit2.http.GET;
import retrofit2.http.Headers;
import retrofit2.http.Query;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        initApiHoder();
    }

    private void initApiHoder() {
        ApiHolderUtil util = new ApiHolderUtil(null, null);
        List<PartApi> parts = new ArrayList<>();
        parts.add(new PartApi(ApiA.class, "https://www.a.com/"));
        parts.add(new PartApi(ApiB.class, "https://www.b.com/"));
        parts.add(new PartApi(ApiC.class, "https://www.c.com/"));

        ApiHolder api = util.createHolderApi(ApiHolder.class, parts, true);
        api.methodA(1);
        api.methodB(1);
        api.methodC(1);
    }

    private interface ApiHolder extends ApiA, ApiB, ApiC {

    }

    private interface ApiA {
        @GET("user/loginWithScanCode")
        Observable<ResponseBody> methodA(@Query("id") int id);
    }

    private interface ApiB {
        @GET("user/loginWithScanCode")
        Observable<ResponseBody> methodB(@Query("id") int id);
    }

    private interface ApiC {
        @GET("user/loginWithScanCode")
        Observable<ResponseBody> methodC(@Query("id") int id);
    }
}

File: app/src/main/java/com/dhy/apiholder/demo/MainActivity.kt
package com.dhy.apiholder.demo

import android.annotation.SuppressLint
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import com.dhy.apiholder.demo.ApiUtil.Companion.api
import com.dhy.apiholder.demo.ApiUtil.Companion.apiUtil
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        btA.setOnClickListener {
            api.methodA(1).subscribe({
                showMsg("onNext")
            }, {
                it.printStackTrace()
                showMsg(it.message ?: "empty error")
            }, {
                showMsg("onComplete")
            })
        }
        btB.setOnClickListener {
            api.methodB(1).subscribe({
                showMsg("onNext")
            }, {
                it.printStackTrace()
                showMsg(it.message ?: "empty error")
            })
        }
        clearLog.setOnClickListener {
            tvLog.text = ""
        }
        //update api when needed
        apiUtil.updateApi(ApiA::class, "https://www.testA.com/")
    }

    @SuppressLint("SetTextI18n")
    fun showMsg(msg: String) {
        tvLog.text = "${tvLog.text}\n$msg"
    }
}
